test/pdo_sqlsrv/pdo_param_output_variants.phpt

```diff
@@ -0,0 +1,140 @@
+    }
+    catch (Exception $e)
+    {
+        //echo "Failed when calling the reverse procedure\n";
+        echo $e->getMessage();
+        echo "\n";
+    }        
+}
+
+function CreateVariantTable($conn, $tableName)
+{
+    try 
+    {
+        $stmt = $conn->exec("CREATE TABLE [$tableName] ([c1_int] int, [c2_variant] sql_variant)");    
+    }
+    catch (Exception $e)
+    {
+        echo "Failed to create a test table\n";
+        echo $e->getMessage();
+    }        
+
+    $tsql = "INSERT INTO [$tableName] ([c1_int], [c2_variant]) VALUES (1, ?)";    
+    
+    $data = "This is to test if sql_variant works with output parameters";
+    
+    $stmt = $conn->prepare($tsql);
+    $result = $stmt->execute(array($data));
+    if (! $result)
+        echo "Failed to insert data\n";
+}
+
+function TestOutputParam($conn, $tableName)
+{
+    // First, create a temporary stored procedure
+    $procName = GetTempProcName('sqlVariant');
+    
+    $spArgs = "@p1 int, @p2 sql_variant OUTPUT";
+    $spCode = "SET @p2 = ( SELECT [c2_variant] FROM $tableName WHERE [c1_int] = @p1 )";
+    
+    $stmt = $conn->exec("CREATE PROC [$procName] ($spArgs) AS BEGIN $spCode END");
+    $stmt = null;
+
+    $callArgs = "?, ?";
+
+    // Data to initialize $callResult variable. This variable should be different from
+    // the inserted data in the table
+    $initData = "A short text";
+    $callResult = $initData;
+
+    try
+    {
+        $stmt = $conn->prepare("{ CALL [$procName] ($callArgs)}");
+        $stmt->bindValue(1, 1);
+        $stmt->bindParam(2, $callResult, PDO::PARAM_STR, 100);
+        $stmt->execute();
+
+
+    }
+    catch (Exception $e)
+    {
+        if(!strcmp($initData, $callResult))
+        {
+            echo "initialized data and result should be the same";
+        }
+        echo $e->getMessage();
+        echo "\n";
+    }        
+}
+
+function RunTest()
+{
+    StartTest("pdo_param_output_variants");
+    try
+    {
+        include("MsSetup.inc");
+        // Connect
+        $conn = new PDO( "sqlsrv:server=$server;Database=$databaseName", $uid, $pwd);
+        $conn->setAttribute( PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION );
+        echo "\n";
+
+        // Test with a simple stored procedure 
+        TestReverse($conn);
+        
+        // Now test with another stored procedure
+        $tableName = GetTempTableName();   
+        CreateVariantTable($conn, $tableName);
+        
+        TestOutputParam($conn, $tableName);
+        
+        $conn = null;
+    }
+    catch (Exception $e)
+    {
+        echo $e->getMessage();
+    }
+    echo "\nDone\n";
+    EndTest("pdo_param_output_variants");
+}
+
+RunTest();
+
+?>
+--EXPECT--
+﻿
+SQLSTATE[22018]: [Microsoft][ODBC Driver 13 for SQL Server][SQL Server]Operand type clash: nvarchar(max) is incompatible with sql_variant
+SQLSTATE[22018]: [Microsoft][ODBC Driver 13 for SQL Server][SQL Server]Operand type clash: nvarchar(max) is incompatible with sql_variant
+
```

Comment:
the initialized data is php string that will fall into sql varchar* type group, in the [core stmt](https://github.com/Microsoft/msphpsql/blob/dev/source/shared/core_stmt.cpp#L546) when binding output parameters the size is assigned to max to avoid silent truncation. so when it is passed to ODBC the type is max, and the server receives the max char type. For error handling in this part we're benefiting from server side errors, as we cannot identify the type of the variant column beforehand (the same reason why we cannot support output parameters), and we can't constraint the error `SQLSTATE[22018]` to only variant, and it might occur when querying other columns than variant.